lib/ocpp/common/websocket/websocket_libwebsockets.cpp

```diff
@@ -675,6 +677,13 @@ bool WebsocketTlsTPM::connect() {
 
     if (!connected) {
         EVLOG_error << "Conn failed, interrupting.";
+
+        // If we timeouted the on_conn_fail was not dispatched, since it did not had the chance
+        if (timeouted && local_data->get_state() != EConnectionState::ERROR) {
```

Comment:
would it be safer to have this code after the check that local_data != nullptr?
e.g.
```
std::shared_ptr<ConnectionData> local_data = conn_data;
if (local_data != nullptr) {
        if (timeouted && local_data->get_state() != EConnectionState::ERROR) {
            EVLOG_error << "Conn failed with timeout, without disconnect dispatch, dispatching manually.";
            on_conn_fail();
        }
        local_data->do_interrupt();
}
```

Reply:
I think there was a problem where local_data had been reset (i.e. nullptr) and local_data->do_interrupt(); was terminating OCPP.

Reply:
Originally the code was conn_data->do_interrupt(); which was a problem because another thread could reset conn_data causing the invalid pointer dereference.
This was solved by taking a copy of conn_data (local_data) however while waiting on the condition variable another thread can change conn_data. After the wait is complete conn_data is re-fetched to ensure that do_interrupt() is called on the most up-to-date information.